### PR TITLE
Add consignment reference

### PR DESCRIPTION
**Pull request by Lholbourn**
These changes add a function within the `consignmentRepository` file to retrieve the next sequence ID from the DB, an additional file/function to call the encoding library to encode a consignment reference, and a change to the `addConsignment` function to utilise both of these new functions. 

I have changed tests affected by these changes and added tests for the function to encode the reference and the function to retrieve the next sequence ID.

### CODE REVIEW

**Comment by Lholbourn** [comment]
The Jenkins build is currently failing as I was working with my local DB when making changes as the DB migration is not yet merged. I'll change this once the DB migration is properly merged and working :+1:

File: src/test/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReferenceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+  "createConsignmentReference" should "create a full consignment reference when given a numerical ID" in {
+    val consignmentReferenceClass = new CreateConsignmentReference
+    val createdYear = 2020
+    val consignmentSequence = 5000
+
+    val result: String = consignmentReferenceClass.createConsignmentReference(createdYear, consignmentSequence)
+
+    result should include ("LBB")
+    result should equal ("TDR-2020-LBB")
```
Comment (suzannehamilton) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReferenceSpec.scala:20]: It would be really great to have tests for different edge cases here. Give me a shout and I'll show you a ScalaTest trick for testing multiple examples without writing lots of repetitive tests.

File: build.sbt [hunk 2]
```diff
@@ -42,7 +42,7 @@ libraryDependencies ++= Seq(
   "io.circe" %% "circe-optics" % circeVersion,
   "io.circe" %% "circe-generic" % circeVersion,
   "io.circe" %% "circe-generic-extras" % circeVersion,
-  "uk.gov.nationalarchives" %% "consignment-api-db" % "0.0.35",
+  "uk.gov.nationalarchives" %% "consignment-api-db" % "0.0.35-SNAPSHOT",
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply build.sbt:45]: It is yeah, I'll change it to the correct version number once the migrration is deployed :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.nationalarchives.tdr.api.model.consignment
+
+import uk.gov.nationalarchives.oci.{Alphabet, BaseCoder, IncludedAlphabet}
+
+class CreateConsignmentReference {
+  def createConsignmentReference(createdDate: Int, consignmentSequenceId: Long): String = {
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala:6]: Good point, I'll make the params/val names more clear!

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.nationalarchives.tdr.api.model.consignment
+
+import uk.gov.nationalarchives.oci.{Alphabet, BaseCoder, IncludedAlphabet}
+
+class CreateConsignmentReference {
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala:5]: Good point! I'll change it to an object, makes using it easier too since I then don't have to have it as a constructor to the `consignmentService` :raised_hands:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  def createConsignmentReference(createdDate: Int, consignmentSequenceId: Long): String = {
+    val baseNumber = 25
+    val encoded = Alphabet.loadAlphabet(Right(IncludedAlphabet.GCRb25)) match {
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala:8]: Ah I hadn't thought about that aspect! I'll change it so it loads outside the method!

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+      case Left(error) => throw new Exception(s"Theres been an error creating the consignment reference: ${error.getMessage}")
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala:9]: Oh, that's a much better way of doing it! I'll change it to that and have a fiddle around with it :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala [hunk 7]
```diff
@@ -29,13 +33,22 @@ class ConsignmentService(
...
-    val consignmentRow = ConsignmentRow(uuidSource.uuid, addConsignmentInput.seriesid, userId, Timestamp.from(timeSource.now))
-    consignmentRepository.addConsignment(consignmentRow).map(row => Consignment(row.consignmentid, row.userid, row.seriesid))
+    val transferStartYearForReference: Int = calendar.get(Calendar.YEAR)
+     consignmentRepository.getNextConsignmentSequence.flatMap(sequence => {
+       val consignmentRef = consignmentReferenceModel.createConsignmentReference(transferStartYearForReference, sequence)
+       val consignmentRow = ConsignmentRow(
+         uuidSource.uuid,
+         addConsignmentInput.seriesid,
+         userId,
+         Timestamp.from(timeSource.now),
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala:43]: That's a much better approach, i'll change it to follow that pattern! :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala [hunk 8]
```diff
@@ -29,13 +33,22 @@ class ConsignmentService(
...
+         consignmentsequence = Option(sequence))
+       consignmentRepository.addConsignment(consignmentRow).map(row => Consignment(row.consignmentid, row.userid, row.seriesid, Option(consignmentRef)))
```
[the remark this answers is not in the document]
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala:45]: The function returns it so that the `consignmentService` function can then use the value when it adds the row to the Consignment table alongside the other Consignment row info. In the `consignmentService` file, line 39 & 46 (I don't know how to link to that line soz :sweat_smile: )

File: src/test/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReferenceSpec.scala (hunk 1 above)
Comment (Lholbourn) [reply src/test/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReferenceSpec.scala:20]: That would be great yeah, thanks! I'll shoot you a message :smile:

File: build.sbt [hunk 9]
```diff
@@ -51,6 +52,8 @@ libraryDependencies ++= Seq(
   "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2",
   "org.jboss.logging" % "jboss-logging" % "3.4.1.Final",
   "com.lightbend.akka" %% "akka-stream-alpakka-slick" % "1.1.2",
+  "org.scala-lang.modules" %% "scala-parser-combinators" % "1.1.2",
```
Comment (suzannehamilton) [inline build.sbt:55]: Is this being used anywhere?

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (suzannehamilton) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/CreateConsignmentReference.scala]: Did you try loading the alphabet outside of the method? It looks like it's still happening inside.

File: build.sbt (hunk 9 above)
Comment (Lholbourn) [reply build.sbt:55]: It's not, I think it's a copy and paste error :sweat: I'll remove it :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.nationalarchives.tdr.api.model.consignment
+
+import java.io.IOException
+
+import uk.gov.nationalarchives.oci.Alphabet.Alphabet
+import uk.gov.nationalarchives.oci.{Alphabet, BaseCoder, IncludedAlphabet}
+
+object ConsignmentReference {
+  val alphabet: Either[IOException, Alphabet] = Alphabet.loadAlphabet(Right(IncludedAlphabet.GCRb25))
```
Comment (Lholbourn) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:9]: @suzannehamilton I'm loading in the alphabet outside the `createConsignmentReference` method here, I think because I've rebased a few times things look a bit jumbled up now. Apologies if it's hard to read! :sweat_smile:

Comment (suzannehamilton) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:9]: Sorry, I didn't explain that very clearly. 😅 The error handling is still happening on every method call, but it's exactly the same every time.

If you move the error handling out of the method, `alphabet` has a much clearer type (`Alphabet`, rather than `Either[IOException, Alphabet]`), and the error handling only happens once. It also has the useful effect that the error handling will happen when the app starts up. This means we'll catch the error on deployment, rather than only when a consignment is created.

Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:9]: Ah I see! I'll change it! :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.nationalarchives.tdr.api.model.consignment
+
+import uk.gov.nationalarchives.oci.Alphabet.Alphabet
+import uk.gov.nationalarchives.oci.{Alphabet, BaseCoder, IncludedAlphabet}
+
+object ConsignmentReference {
+  val alphabet: Alphabet = Alphabet.loadAlphabet(Right(IncludedAlphabet.GCRb25)) match {
+    case Left(error) => throw new Exception(s"Theres been an error creating the consignment reference", error)
```
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:8]: Just an update to the error message: it happens when the alphabet is being loaded rather than when the consignment reference is being created.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala [hunk 12]
```diff
@@ -29,13 +33,23 @@ class ConsignmentService(
   }
 
   def addConsignment(addConsignmentInput: AddConsignmentInput, userId: UUID): Future[Consignment] = {
-    val consignmentRow = ConsignmentRow(uuidSource.uuid, addConsignmentInput.seriesid, userId, Timestamp.from(timeSource.now))
-    consignmentRepository.addConsignment(consignmentRow).map(row => Consignment(row.consignmentid, row.userid, row.seriesid))
+    val now = timeSource.now
+    val yearNow = LocalDate.from(now.atOffset(ZoneOffset.UTC)).getYear
+     consignmentRepository.getNextConsignmentSequence.flatMap(sequence => {
```
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala:38]: Something odd has happened to the indentation here.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    case Right(alphabet) => alphabet
+  }
+
+  def createConsignmentReference(transferStartYear: Int, consignmentSequenceId: Long): String = {
+    val baseNumber = 25
```
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:13]: Could this value also be a constant, defined outside of the method? Just to make it easier to distinguish the parts that never change from the dynamic stuff.

Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:13]: Ah yeah, didn't think of that. I'll change that :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala (hunk 12 above)
Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentService.scala:38]: Oh yeah, I'll fix it! :+1:

**Comment by suzannehamilton** [comment]
Worth rebasing on the master branch. It looks like you have some merge conflicts, and the fix for the flaky test has just been merged.

**Comment by Lholbourn** [comment]
> Worth rebasing on the master branch. It looks like you have some merge conflicts, and the fix for the flaky test has just been merged.

Will do! In the process of picking through some conflicts now :+1:

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala (hunk 11 above)
Comment (suzannehamilton) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:8]: Sorry, I think I accidentally resolved this conversation rather than one of the other ones.

**Comment by suzannehamilton** [comment]
Look good to me. I think there's still just one open comment about the error message wording: https://github.com/nationalarchives/tdr-consignment-api/pull/108#pullrequestreview-581133068

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.nationalarchives.tdr.api.model.consignment
+
+import uk.gov.nationalarchives.oci.Alphabet.Alphabet
+import uk.gov.nationalarchives.oci.{Alphabet, BaseCoder, IncludedAlphabet}
+
+object ConsignmentReference {
+  val alphabet: Alphabet = Alphabet.loadAlphabet(Right(IncludedAlphabet.GCRb25)) match {
+    case Left(error) => throw new Exception(s"Theres been an error loading the encoding alphabet", error)
```
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:8]: Sorry, one last thing: missing apostrophe in "Theres".

You could just drop the "Theres been an" part of it anyway, and say something like: `Error loading the GCRb25 encoding alphabet`.

Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:8]: Sure, will do! :+1:

Comment (suzannehamilton) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala:8]: Is there a commit that still needs to be pushed? This message looks the same as before.

Comment (Lholbourn) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/consignment/ConsignmentReference.scala]: Ah weird, I think GitHub had a bit of a tizzy, I pushed it yesterday but it's only showing up now. I'll need to fix some conflicts now anyway though.

**Review by suzannehamilton — APPROVED** [review]
🎉
